### ISSUE TITLE
Training loop update, unexpected behaviour

### ISSUE THREAD

**Issue by PaoDuroMacMoney** [explosion/spaCy#2605]
## How to reproduce the behaviour
In the documentation, we use the following code for training a model:

```
    for text, annotations in TRAIN_DATA:
        nlp.update([text], [annotations], sgd=optimizer, drop=0.35, losses=losses)
```
In Language's update method, the text list and annotation lists are transformed into a doc list and a GoldenParser list. The update propagates theses objs lists throught pipe's components update methods. So, if instead of call update each time for each doc, we could pass the array of texts and annotations directly to the method:

```
    texts = [text for text,_ in TRAIN_DATA]
    annotations = [annotation for _,annotation in TRAIN_DATA]
    nlp.update(texts , annotations, sgd=optimizer, drop=0.35, losses=losses)
```
However in this form of calling the update, the iterations over a batch runs faster, but loss values are quite smaller. 

I'm not really sure this is a bug or this is really the expected behaviour, 

Thanks in advance.

## Your Environment

* **spaCy version:** 2.0.11
* **Platform:** Windows-10-10.0.16299-SP0
* **Python version:** 3.6.5

**Comment by honnibal** [comment]
It's hard to tell what's going wrong in your example, because I don't have the full loop there. But it looks to me like you're creating the `doc` and `gold` objects once, so maybe they're being modified? Or maybe you've batched the data incorrectly? I'm not sure, but I suspect the bug is in your code, rather than the update method.

**Comment by PaoDuroMacMoney** [comment]
Ok, I'm posting full code example based on training loop from documentation, with both versions and outputs. The 2 main differences I see is time of execution of iterations and losses that operates in a different range:

Version 1 - For loop just like documentation:

```
#!/usr/bin/env python
# coding: utf-8

from __future__ import unicode_literals, print_function

import plac
import random
import spacy
import time

TRAIN_DATA = [
    ("find a cafe with great wifi", {
        'heads': [0, 2, 0, 5, 5, 2],  # index of token head
        'deps': ['ROOT', '-', 'PLACE', '-', 'QUALITY', 'ATTRIBUTE']
    }),
    ("find a hotel near the beach", {
        'heads': [0, 2, 0, 5, 5, 2],
        'deps': ['ROOT', '-', 'PLACE', 'QUALITY', '-', 'ATTRIBUTE']
    }),
    ("find me the closest gym that's open late", {
        'heads': [0, 0, 4, 4, 0, 6, 4, 6, 6],
        'deps': ['ROOT', '-', '-', 'QUALITY', 'PLACE', '-', '-', 'ATTRIBUTE', 'TIME']
    }),
    ("show me the cheapest store that sells flowers", {
        'heads': [0, 0, 4, 4, 0, 4, 4, 4],  # attach "flowers" to store!
        'deps': ['ROOT', '-', '-', 'QUALITY', 'PLACE', '-', '-', 'PRODUCT']
    }),
    ("find a nice restaurant in london", {
        'heads': [0, 3, 3, 0, 3, 3],
        'deps': ['ROOT', '-', 'QUALITY', 'PLACE', '-', 'LOCATION']
    }),
    ("show me the coolest hostel in berlin", {
        'heads': [0, 0, 4, 4, 0, 4, 4],
        'deps': ['ROOT', '-', '-', 'QUALITY', 'PLACE', '-', 'LOCATION']
    }),
    ("find a good italian restaurant near work", {
        'heads': [0, 4, 4, 4, 0, 4, 5],
        'deps': ['ROOT', '-', 'QUALITY', 'ATTRIBUTE', 'PLACE', 'ATTRIBUTE', 'LOCATION']
    })
]

@plac.annotations(n_iter=("Number of training iterations", "option", "n", int))
def main(n_iter=20):
    nlp = spacy.blank('en')
    print("Created blank 'en' model")

    if 'parser' in nlp.pipe_names:
        nlp.remove_pipe('parser')
    parser = nlp.create_pipe('parser')
    nlp.add_pipe(parser, first=True)

    for text, annotations in TRAIN_DATA:
        for dep in annotations.get('deps', []):
            parser.add_label(dep)

    other_pipes = [pipe for pipe in nlp.pipe_names if pipe != 'parser']
    with nlp.disable_pipes(*other_pipes):  # only train parser
        optimizer = nlp.begin_training()
        for itn in range(n_iter):
            start = time.time()
            random.shuffle(TRAIN_DATA)
            losses = {}
            for text, annotations in TRAIN_DATA:
                nlp.update([text], [annotations], sgd=optimizer, losses=losses)
            duration = time.time() - start
            print('{} - {:.3f} seconds '.format(losses, duration))

if __name__ == '__main__':
    plac.call(main)
```
Outputs:

```
{'parser': 48.516792081637654} - 1.163 seconds 
{'parser': 34.55893930891463} - 1.077 seconds 
{'parser': 24.381089747448257} - 1.053 seconds 
{'parser': 18.189276725090266} - 1.088 seconds 
{'parser': 8.978527705645075} - 1.086 seconds 
{'parser': 7.313750271940936} - 1.101 seconds 
{'parser': 7.7396583363407485} - 1.058 seconds 
{'parser': 1.2773729733026489} - 1.095 seconds 
{'parser': 2.802887417285215} - 1.066 seconds 
{'parser': 0.0005350369869161497} - 1.084 seconds 
{'parser': 3.9477350383060053} - 1.059 seconds 
{'parser': 1.4194515017761364} - 1.066 seconds 
{'parser': 1.836798675903791} - 1.073 seconds 
{'parser': 3.9330670102429792} - 1.057 seconds 
{'parser': 2.0014526367719636} - 1.062 seconds 
{'parser': 7.858103645115585e-05} - 1.079 seconds 
{'parser': 1.793029156003163e-08} - 1.081 seconds 
{'parser': 5.3666862959720335e-09} - 1.081 seconds 
{'parser': 2.4006624866697287e-09} - 1.070 seconds 
{'parser': 1.1688846660489816e-09} - 1.071 seconds 
```
Version 2 - Using list of texts and annotations on update call:

```
#!/usr/bin/env python
# coding: utf-8

from __future__ import unicode_literals, print_function

import plac
import random
import spacy
import time

TRAIN_DATA = [
    ("find a cafe with great wifi", {
        'heads': [0, 2, 0, 5, 5, 2],  # index of token head
        'deps': ['ROOT', '-', 'PLACE', '-', 'QUALITY', 'ATTRIBUTE']
    }),
    ("find a hotel near the beach", {
        'heads': [0, 2, 0, 5, 5, 2],
        'deps': ['ROOT', '-', 'PLACE', 'QUALITY', '-', 'ATTRIBUTE']
    }),
    ("find me the closest gym that's open late", {
        'heads': [0, 0, 4, 4, 0, 6, 4, 6, 6],
        'deps': ['ROOT', '-', '-', 'QUALITY', 'PLACE', '-', '-', 'ATTRIBUTE', 'TIME']
    }),
    ("show me the cheapest store that sells flowers", {
        'heads': [0, 0, 4, 4, 0, 4, 4, 4],  # attach "flowers" to store!
        'deps': ['ROOT', '-', '-', 'QUALITY', 'PLACE', '-', '-', 'PRODUCT']
    }),
    ("find a nice restaurant in london", {
        'heads': [0, 3, 3, 0, 3, 3],
        'deps': ['ROOT', '-', 'QUALITY', 'PLACE', '-', 'LOCATION']
    }),
    ("show me the coolest hostel in berlin", {
        'heads': [0, 0, 4, 4, 0, 4, 4],
        'deps': ['ROOT', '-', '-', 'QUALITY', 'PLACE', '-', 'LOCATION']
    }),
    ("find a good italian restaurant near work", {
        'heads': [0, 4, 4, 4, 0, 4, 5],
        'deps': ['ROOT', '-', 'QUALITY', 'ATTRIBUTE', 'PLACE', 'ATTRIBUTE', 'LOCATION']
    })
]


@plac.annotations(n_iter=("Number of training iterations", "option", "n", int))
def main(n_iter=20):
    nlp = spacy.blank('en')
    print("Created blank 'en' model")

    if 'parser' in nlp.pipe_names:
        nlp.remove_pipe('parser')
    parser = nlp.create_pipe('parser')
    nlp.add_pipe(parser, first=True)

    for text, annotations in TRAIN_DATA:
        for dep in annotations.get('deps', []):
            parser.add_label(dep)

    other_pipes = [pipe for pipe in nlp.pipe_names if pipe != 'parser']
    with nlp.disable_pipes(*other_pipes):  # only train parser
        optimizer = nlp.begin_training()
        for itn in range(n_iter):
            start = time.time()
            random.shuffle(TRAIN_DATA)
            losses = {}
            nlp.update([text for text, _ in TRAIN_DATA], [annotations for _, annotations in TRAIN_DATA], sgd=optimizer, losses=losses)
            duration = time.time() - start
            print('{} - {:.3f} seconds '.format(losses, duration))

if __name__ == '__main__':
    plac.call(main)
```
Outputs:

```
{'parser': 1.1612772420048714} - 0.234 seconds 
{'parser': 0.6587301231920719} - 0.162 seconds 
{'parser': 0.3522218959406018} - 0.172 seconds 
{'parser': 0.1260265299169987} - 0.156 seconds 
{'parser': 0.06156431674162377} - 0.156 seconds 
{'parser': 0.05281189433117106} - 0.156 seconds 
{'parser': 0.00488906922885235} - 0.156 seconds 
{'parser': 1.9246498925706845e-05} - 0.156 seconds 
{'parser': 1.3216422968115316e-06} - 0.165 seconds 
{'parser': 3.221937706410216e-06} - 0.156 seconds 
{'parser': 3.830143411320819e-07} - 0.156 seconds 
{'parser': 2.336446253033142e-06} - 0.172 seconds 
{'parser': 2.84883725512608e-07} - 0.172 seconds 
{'parser': 6.719809165708912e-08} - 0.172 seconds 
{'parser': 2.2236903252786933e-08} - 0.162 seconds 
{'parser': 1.1845976234658995e-08} - 0.156 seconds 
{'parser': 7.876341183432611e-09} - 0.172 seconds 
{'parser': 5.571918357261119e-09} - 0.156 seconds 
{'parser': 4.3459764936110146e-09} - 0.156 seconds 
{'parser': 3.890155494176012e-09} - 0.156 seconds 
```
The same goes for NER training

**Comment by honnibal** [comment]
Yeah in your second example you're making one batch of updates, in the first example you're making a series of updates, each with a batch of one example. These aren't the same thing.

Each call to `.update()` results in a modification to the weights, based on a gradient estimated from a minibatch of examples. If you call `.update()` with all your data, you make one gradient estimate, and perform one update. If you call it with one item of your data, you estimate the gradient based on a smaller sample (so the gradient estimate is noisier), and then make the change to the weights.

Note that the noise in the gradient estimates is actually by design in stochastic gradient descent --- that's the "stochastic" part. It adds an element of exploration to the algorithm, making it less sensitive to initialization. So, you can expect using the right batch size to lead to better generalization, not just faster convergence. The interaction between batch size, wall clock time, generalization and training loss is complicated, and a subject of ongoing research.
